Source/WebKit/UIProcess/Launcher/glib/BubblewrapLauncher.cpp

```diff
@@ -686,16 +716,31 @@ GRefPtr<GSubprocess> bubblewrapSpawn(GSubprocessLauncher* launcher, const Proces
         "--ro-bind-try", PKGLIBEXECDIR, PKGLIBEXECDIR,
     };
 
+    addExtraPaths(launchOptions.extraSandboxPaths, sandboxArgs);
+
     if (launchOptions.processType == ProcessLauncher::ProcessType::DBusProxy) {
         sandboxArgs.appendVector(Vector<CString>({
             "--ro-bind", DBUS_PROXY_EXECUTABLE, DBUS_PROXY_EXECUTABLE,
-            // This is a lot of access, but xdg-dbus-proxy is trusted so that's OK. It's sandboxed
-            // only because we have to mount .flatpak-info in its mount namespace. The user rundir
-            // is where we mount our proxy socket.
-            "--bind", runDir, runDir,
+            "--bind", dbusProxyDirectory(), dbusProxyDirectory(),
         }));
 
-        addExtraPaths(launchOptions.extraSandboxPaths, sandboxArgs);
+        // xdg-dbus-proxy is trusted, so it's OK to mount the directories that contain the session
+        // bus and a11y bus sockets wherever they may be. xdg-dbus-proxy is sandboxed only because
+        // we have to mount .flatpak-info in its mount namespace so that portals may use it as a
+        // trusted way to get the app ID of the process that is using it.
+        if (auto sessionBusDirectory = directoryContainingDBusSocket(g_getenv("DBUS_SESSION_BUS_ADDRESS"))) {
+            sandboxArgs.appendVector(Vector<CString>({
+                "--bind", *sessionBusDirectory, *sessionBusDirectory,
+            }));
+        }
+
+#if ENABLE(ACCESSIBILITY)
+        if (auto a11yBusDirectory = directoryContainingDBusSocket(PlatformDisplay::sharedDisplay().computeAccessibilityBusAddress().utf8().data())) {
+            sandboxArgs.appendVector(Vector<CString>({
+                "--bind", *a11yBusDirectory, *a11yBusDirectory,
+            }));
+        }
+#endif
```

Comment:
The problem of this approach is that in the case of wayland we end up doing two sync dbus calls to get the actual a11y bus address, once here in and one in XDGDBusProxy::accessibilityProxy(). So, I think using the PlatformDisplay to override the a11y bus address was not a good idea, because it's only done in the UI process which is not sandboxed, but only because we need a way to pass the sandboxed bus address to the web process via creation parameters. What I would do is:
- Remove the setAccessibilityBusAddress from PlatformDisplay: the display always returns the original bus address and it's cached to ensure there's only one sync dbus call.
- Add setSandboxedAccessibilityBusAddress() and sandboxedAccessibilityBusAddress() to Sandbox.h in WTF
- Use WTF Sandboxed API to set and get the global sandboxed a11y bus in the UI process.

What do you think?

Reply:
OK, that ought to work.

Reply:
I've implemented your suggestion. Does it look OK?